pkg/kube/runnable_job.go

```diff
@@ -66,30 +61,23 @@ func (r *runnableJob) Run(ctx context.Context) error {
 
 	var err error
 
-	for i, s := range r.secrets {
-		klog.V(3).Infof("Creating secret %q", r.job.Namespace+"/"+s.Name)
-		r.secrets[i], err = r.clientset.CoreV1().Secrets(r.job.Namespace).Create(ctx, s, metav1.CreateOptions{})
-		if err != nil {
-			return err
-		}
-	}
-
 	klog.V(3).Infof("Creating job: %s/%s", r.job.Namespace, r.job.Name)
 	r.job, err = r.clientset.BatchV1().Jobs(r.job.Namespace).Create(ctx, r.job, metav1.CreateOptions{})
 	if err != nil {
 		return fmt.Errorf("creating job: %w", err)
 	}
 
-	for i, s := range r.secrets {
-		klog.V(3).Infof("Setting owner reference for secret %q", r.job.Namespace+"/"+s.Name)
-		err = controllerutil.SetOwnerReference(r.job, s, r.scheme)
+	for i, secret := range r.secrets {
+		secret.Namespace = r.job.Namespace
+		klog.V(3).Infof("Setting owner reference for secret %q", r.job.Namespace+"/"+secret.Name)
+		err = controllerutil.SetOwnerReference(r.job, secret, r.scheme)
 		if err != nil {
 			return err
 		}
-		klog.V(3).Infof("Updating secret %q", r.job.Namespace+"/"+s.Name)
-		r.secrets[i], err = r.clientset.CoreV1().Secrets(r.job.Namespace).Update(ctx, s, metav1.UpdateOptions{})
+		klog.V(3).Infof("Creating secret %q", r.job.Namespace+"/"+secret.Name)
+		r.secrets[i], err = r.clientset.CoreV1().Secrets(r.job.Namespace).Create(ctx, secret, metav1.CreateOptions{})
```

Comment:
For my understanding: why does it work even though `Secret` is created after `Job` starts? IIUC, `Job` loads those secrets in the spec as environmental variables. I thought they should exist before Job starts.

Reply:
Very good question. I think it works because a Job does not required secrets to be present when it's created. Only the pod that it controls requires those secrets to be present. Since it takes some time for Job controller to pick it up and create the pod, by that time the secrets are also created.

So I tested it and apparently works, but now I'm wondering if it might lead to some errors when a pod is scheduled very very fast.

Reply:
Yes, I'm worried about that. It may also take longer for some other reason to create `Secret`. `Job` might create the Pod before creating `Secret`. It works in many cases as you said, but it might cause an unclear error occasionally.

Reply:
Okey let me revert that for the CLI and do the same for the operator:

- Create Secrets referred to by the Job
- Create the Job
- Update Secrets' ownerReferences to point to the Job

This is an extra Kubernetes API cal but more readable and predictable.

Reply:
I'm not sure it is possible to happen. If you can say it never happens, you don't have to change the order. I'm just concerned.

Reply:
I think it may happen even though it's not very likely. Anyway I've changed that to order to first create secrets and then job.

Reply:
I've already approved :shipit: